lib/spack/spack/solver/asp.py

```diff
@@ -774,6 +774,38 @@ def package_compiler_defaults(self, pkg):
                 pkg.name, cspec.name, cspec.version, -i * 100
             ))
 
+    def cfg_required_rules(self, pkg):
+        experimental_hack_selector = 1
+        if experimental_hack_selector == 1:
+            # The goal here is to say "if we need to build MPI, then it has to
+            # either satisfy @4.0.2+static or @4.0.1~static"
+            test_rules = [
+                ("openmpi",
+                 # This list of specs is an example of a stanza
+                 ["@4.0.2+static",  # Example stanza member
+                  "@4.0.1~static"]
+                )
+            ]
+        elif experimental_hack_selector == 2:
+            # Example of combinatoric constraints, you can mix either compiler
+            # with either version, but it must be one of the 4 combinations
+            test_rules = [
+                ("openmpi", ["%gcc", "%clang"]),
+                ("openmpi", ["@4.0.1", "@4.0.0"])
+            ]
```

Comment:
Proposed schema for `packages.yaml`:

```
packages:
  openmpi:
    spec_requirements:
    - require_one_of:
      - @4.0.2+static
      - @4.0.1~static
```

and

```
packages:
  openmpi:
    spec_requirements:
    - require_one_of:
      - %gcc
      - %clang
    - require_one_of:
      - @4.0.1
      - @4.0.2
    - require: +static # shorthand when you always want to enforce a given spec
```

(these correspond to the examples currently in asp.py)